zio/parquetio/vectorreader.go

```diff
@@ -0,0 +1,345 @@
+package parquetio
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/memory"
+	"github.com/apache/arrow-go/v18/parquet"
+	"github.com/apache/arrow-go/v18/parquet/file"
+	"github.com/apache/arrow-go/v18/parquet/pqarrow"
+	"github.com/brimdata/super"
+	"github.com/brimdata/super/pkg/field"
+	"github.com/brimdata/super/vector"
+	"github.com/brimdata/super/zio/arrowio"
+)
+
+type VectorReader struct {
+	ctx  context.Context
+	zctx *super.Context
+
+	fr           *pqarrow.FileReader
+	colIndexes   []int
+	nextRowGroup *atomic.Int64
+	rr           pqarrow.RecordReader
+	vb           vectorBuilder
+}
+
+func NewVectorReader(ctx context.Context, zctx *super.Context, r io.Reader, fields []field.Path) (*VectorReader, error) {
+	ras, ok := r.(parquet.ReaderAtSeeker)
+	if !ok {
+		return nil, errors.New("reader cannot seek")
+	}
+	pr, err := file.NewParquetReader(ras)
+	if err != nil {
+		return nil, err
+	}
+	pqprops := pqarrow.ArrowReadProperties{
+		Parallel:  true,
+		BatchSize: 16184,
+	}
+	fr, err := pqarrow.NewFileReader(pr, pqprops, memory.NewGoAllocator())
+	if err != nil {
+
+	}
+	return &VectorReader{
+		ctx:          ctx,
+		zctx:         zctx,
+		fr:           fr,
+		colIndexes:   columnIndexes(fr.Manifest, fields),
+		nextRowGroup: &atomic.Int64{},
+		vb:           vectorBuilder{zctx, map[arrow.DataType]super.Type{}},
+	}, nil
+}
+
+func (p *VectorReader) NewConcurrentPuller() vector.Puller {
+	return &VectorReader{
+		ctx:          p.ctx,
+		fr:           p.fr,
+		colIndexes:   p.colIndexes,
+		nextRowGroup: p.nextRowGroup,
+		vb:           vectorBuilder{p.zctx, map[arrow.DataType]super.Type{}},
+	}
+}
+func (p *VectorReader) Pull(done bool) (vector.Any, error) {
+	if done {
+		return nil, nil
+	}
+	for {
+		if err := p.ctx.Err(); err != nil {
+			return nil, err
+		}
+		if p.rr == nil {
+			rowGroup := int(p.nextRowGroup.Add(1) - 1)
+			if rowGroup >= p.fr.ParquetReader().NumRowGroups() {
+				return nil, nil
+			}
+			rr, err := p.fr.GetRecordReader(p.ctx, p.colIndexes, []int{rowGroup})
+			if err != nil {
+				return nil, err
+			}
+			p.rr = rr
+		}
+		rec, err := p.rr.Read()
+		if err != nil {
+			if err == io.EOF {
+				p.rr = nil
+				continue
+			}
+			return nil, err
+		}
+		return p.vb.build(array.RecordToStructArray(rec))
+	}
+}
+
+type vectorBuilder struct {
+	zctx  *super.Context
+	types map[arrow.DataType]super.Type
+}
+
+func (v *vectorBuilder) build(a arrow.Array) (vector.Any, error) {
+	dt := a.DataType()
+	length := uint32(a.Len())
+	// Order here follows that of the arrow.Type constants.
+	switch dt.ID() {
+	case arrow.NULL:
+		bits := make([]uint64, (length+7)/8)
+		for i := range bits {
+			bits[i] = ^uint64(0)
+		}
+		nulls := vector.NewBool(bits, length, nil)
+		return vector.NewConst(super.Null, length, nulls), nil
```

Comment:
Why would a null const value have a nulls vector? Is there such thing as a null null?

Reply:
To the vector runtime, a slot in a vector is null if and only if that vector has a nulls vector and its bit corresponding to that slot is set. A vector.Const containing a null zed.Value but having no nulls vector is generally going to look to the runtime like it contains zero values rather than nulls.

We could of course change the runtime to check for a null slot by first checking the nulls vector and then falling back to checking for a vector.Const containing a null zed.Value, but my take is that's more complicated than the current approach without offering any benefit.

This is fairly subtle so my explanation here might not be very convincing, in which case I'm happy to elaborate.